effectful/handlers/torch.py

```diff
@@ -180,20 +170,70 @@ def reindex_flat_tensor(t):
             return t
 
         result = t.reshape(inds[0].shape + t.shape[1:])
-        return torch_getitem(result, tuple(var() for (var, _) in reindex_fvs))
+        return torch_getitem(result, tuple(k() for k in sized_fvs.keys()))
+
+    result = tree.map_structure(reindex_flat_tensor, flat_result)
+    return result
+
+
+@defop
+def to_tensor_(
+    t: Annotated[torch.Tensor, Scoped[A | B]],
+    *args: Annotated[Operation[[], torch.Tensor], Scoped[A]],
+) -> Annotated[torch.Tensor, Scoped[B]]:
+    def _evaluate(expr):
+        if isinstance(expr, Term):
+            (args, kwargs) = tree.map_structure(_evaluate, (expr.args, expr.kwargs))
+            return _partial_eval(expr)
+        if tree.is_nested(expr):
+            return tree.map_structure(_evaluate, expr)
+        return expr
+
+    if not isinstance(t, Term):
+        return t
 
-    return tree.map_structure(reindex_flat_tensor, flat_result)
+    result = _evaluate(t)
+    if not isinstance(result, Term) or not args:
+        return result
+
+    # ensure that the result is a torch_getitem with a tensor as the first argument
+    if not (result.op is torch_getitem and isinstance(result.args[0], torch.Tensor)):
+        raise NotImplementedError
+
+    tensor = result.args[0]
+    dims = result.args[1]
+    assert isinstance(dims, Sequence)
+
+    # ensure that the order is a subset of the named dimensions
+    order_set = set(args)
+    if not order_set <= set(a.op for a in dims if isinstance(a, Term)):
+        raise NotImplementedError
+
+    # permute the inner tensor so that the leading dimensions are in the order
+    # specified and the trailing dimensions are the remaining named dimensions
+    # (or slices)
+    reindex_dims = [
+        i
+        for i, o in enumerate(dims)
+        if not isinstance(o, Term) or o.op not in order_set
+    ]
+    dim_ops = [a.op if isinstance(a, Term) else None for a in dims]
+    perm = [dim_ops.index(o) for o in args] + reindex_dims
+    tensor = tensor.permute(perm)
+    return tensor[(slice(None),) * len(args) + tuple(dims[i] for i in reindex_dims)]
```

Comment:
I'm not seeing the equivalent code in the previous version for this block at the end of `to_tensor_`. What is this logic doing that is not accomplished by `_partial_eval`'s `reindex_flat_tensor`? Why is the `permute` call above necessary? Is this meant to fix a bug?

Reply:
The distinction between the two functions is that `_partial_eval` is only concerned with the partial evaluation rule and `to_tensor` applies partial evaluation, then reorders dimensions. In the previous implementation, `_partial_eval` did both things and the permute was implicit in `ordered_sized_fvs`.